File: app/src/main/java/nl/giorgos/appie/inject/AppComponent.java
package nl.giorgos.appie.inject;

import javax.inject.Singleton;

import dagger.Component;
import nl.giorgos.appie.AppieApp;
import nl.giorgos.appie.base.BaseActivity;
import nl.giorgos.appie.category.CategoryActivity;
import nl.giorgos.appie.country_news_list.CountryNewsListActivity;
import nl.giorgos.appie.news_detail.NewsDetailActivity;

@Singleton
@Component(modules = {
        AppModule.class,
        NetworkModule.class
})
public interface AppComponent {

    void inject(AppieApp app);

    void inject(BaseActivity baseActivity);

    void inject(CategoryActivity categoryActivity);

    void inject(CountryNewsListActivity countryNewsListActivity);

    void inject(NewsDetailActivity newsDetailActivity);
}

File: app/src/main/java/nl/giorgos/launchme/inject/AppComponent.kt
package nl.giorgos.launchme.inject

import android.app.Application
import androidx.lifecycle.ViewModelProvider
import com.google.gson.GsonBuilder
import dagger.Component
import dagger.Module
import dagger.Provides
import nl.giorgos.launchme.launch.api.BaseUrl
import nl.giorgos.launchme.launch.api.LaunchExecutor
import nl.giorgos.launchme.launch.api.LaunchRepository
import nl.giorgos.launchme.launch.api.LaunchService
import nl.giorgos.launchme.launch.detail.DetailActivity
import nl.giorgos.launchme.launch.detail.DetailViewModel
import nl.giorgos.launchme.launch.list.LaunchListActivity
import nl.giorgos.launchme.launch.list.LaunchListViewModel
import nl.giorgos.launchme.launch.utils.ViewModelsRepository
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
open class Modules(private val application: Application) {
    @Provides
    @Singleton
    open fun providesRetrofit(): Retrofit {
        //20190222T014500Z
        val gson = GsonBuilder()
            .setDateFormat("yyyyMMdd'T'HHmmss")
            .create()
        return Retrofit.Builder().baseUrl(BaseUrl)
            .addConverterFactory(GsonConverterFactory.create(gson)).build()
    }

    @Provides
    open fun providesLaunchService(retrofit: Retrofit): LaunchService {
        return retrofit.create(LaunchService::class.java)
    }

    @Provides
    @Singleton
    open fun providesLaunchRepository(launchExecutor: LaunchExecutor): LaunchRepository {
        return LaunchRepository(launchExecutor)
    }
}

@Singleton
@Component(modules = arrayOf(Modules::class))
interface AppComponent {
    fun inject(launchListActivity: LaunchListActivity)
    fun inject(launchListViewModel: LaunchListViewModel)

    fun inject(detailActivity: DetailActivity)
    fun inject(detailViewModel: DetailViewModel)
}